test/varmat_compatibility.py

```diff
@@ -0,0 +1,260 @@
+    associated with all the functions in functions_or_sigs, or if functions_or_sigs
+    is empty every signature the stanc3 compiler exposes.
+
+    Results are written to a results json file. Individual signatures are classified
+    as either compatible, incompatible, or impossible
+
+    Compatible signatures can be compiled with varmat types in every argument that
+    could possibly be a varmat (the matrix-like ones).
+
+    Incompatible signatures cannot all be built, and for impossible signatures it does
+    not make sense to try to build them (there are no matrix arguments, or the function
+    does not support reverse mode autodiff, etc)
+
+    Compilation is done in parallel using the number of specified cores
+
+    :param functions_or_sigs: List of function names and/or signatures to benchmark
+    :param results_file: File to use as a results cache
+    :param cores: Number of cores to use for compiling
+    """
+    all_signatures = get_signatures()
+    functions, signatures = handle_function_list(functions_or_sigs)
+
+    requested_functions = set(functions)
+
+    compatible_signatures = set()
+    incompatible_signatures = set()
+    impossible_signatures = set()
+
+    # Read the arguments and figure out the exact list of signatures to test
+    signatures_to_check = set()
+    for signature in all_signatures:
+        sp = SignatureParser(signature)
+
+        if len(requested_functions) > 0 and sp.function_name not in requested_functions:
+            continue
+        
+        signatures_to_check.add(signature)
+
+    work_queue = Queue.Queue()
+
+    # For each signature, generate cpp code to test
+    for n, signature in enumerate(signatures_to_check):
+        sp = SignatureParser(signature)
+
+        if sp.is_high_order():
+            work_queue.put((n, signature, None))
+            continue
+
+        cpp_code = ""
+        any_overload_uses_varmat = False
+        
+        for m, overloads in enumerate(itertools.product(("Prim", "Rev", "RevVarmat"), repeat = sp.number_arguments())):
+            cg = CodeGenerator()
+
+            arg_list_base = cg.build_arguments(sp, overloads, size = 1)
+
+            arg_list = []
+            for overload, arg in zip(overloads, arg_list_base):
+                if arg.is_reverse_mode() and arg.is_varmat_compatible() and overload.endswith("Varmat"):
+                    any_overload_uses_varmat = True
+                    arg = cg.to_var_value(arg)
+                
+                arg_list.append(arg)
+
+            cg.function_call_assign("stan::math::" + sp.function_name, *arg_list)
+
+            cpp_code += BENCHMARK_TEMPLATE.format(
+                benchmark_name = sp.function_name + repr(m),
+                code=cg.cpp(),
+            )
+
+        if any_overload_uses_varmat:
+            work_queue.put((n, signature, cpp_code))
+        else:
+            work_queue.put((n, signature, None))
```

Comment:
Why do we even need non varmat functions in work queue?

Reply:
Good Q. Will think about this more.

Reply:
I added a comment.

Reply:
Would not this be simpler if you just immediately added these into irrelevant set?

Reply:
> Would not this be simpler if you just immediately added these into irrelevant set?

I was thinking having the output print in all the same place, but if at the start we just printed all the irrelevant things, that seems better.

Reply:
I switched to this and it is indeed better.

Reply:
Output looks like:

```
sum(array[] int) => int ... Irrelevant
sum(array[] real) => real ... Irrelevant
Results of test 0 / 3, sum(row_vector) => real ... Success
Results of test 1 / 3, sum(matrix) => real ... Success
Results of test 2 / 3, sum(vector) => real ... Success
```